### PR TITLE
183 | add a regression test for dnatracing against minicircle.spm

### PR DESCRIPTION
**Pull request by ns-rse**
Before starting work on refactoring `dnatracing.py` I've added a "regression test" which checks what the overall output of the process should be and am putting it out for PR/review so that they are smaller and easier to review/understand.

The purpose is that as I go about refactoring the code I run this test, if it fails I've made a mistake and have incorrectly changed the process.

### CODE REVIEW

**Comment by ns-rse** [comment]
Of course it would help if I didn't submit mistakes wouldn't it! Sorry, will correct now.

I've also just realised that I face some problems because I can't use `tmpdir` as the output directory that is stored in the `Basename` column of the CSV since that is going to change depending on where/when the test is run. :thinking:

**Comment by ns-rse** [comment]
That took longer than I thought it would to resolve but tests now pass here on GitHub. Apologies for the "noise".

File: tests/test_dnatracing.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
+"""Tests of dnatracing methods."""
+import numpy as np
+import pytest
+
+
+def test_get_grain_array(test_dnatracing) -> None:
+    """Test extraction of grain array."""
+    TARGET = {
```
Comment (bobturneruk) [inline tests/test_dnatracing.py:8]: Would this be better as a fixture? If only to future-proof?

File: tests/conftest.py [hunk 2]
```diff
@@ -397,3 +398,48 @@ def minicircle_grainstats(
 def minicircle_grainstats_20220526() -> pd.DataFrame:
     """Statistics for minicircle for comparison."""
     return pd.read_csv(RESOURCES / "minicircle_grainstats_20220526.csv", index_col=0)
+
+
+# Derive fixtures for DNA Tracing
+GRAINS = np.array(
```
Comment (bobturneruk) [inline tests/conftest.py:404]: Should this be a fixture itself? Maybe not.

File: tests/conftest.py [hunk 3]
```diff
@@ -397,3 +398,48 @@ def minicircle_grainstats(
...
+        [0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 2],
+        [0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 2],
+        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2],
+        [0, 0, 3, 3, 3, 3, 3, 0, 0, 0, 2],
+        [0, 0, 3, 0, 0, 0, 3, 0, 0, 0, 2],
+        [0, 0, 3, 3, 3, 3, 3, 0, 0, 0, 2],
+        [0, 0, 4, 4, 4, 4, 4, 0, 0, 0, 2],
+    ]
+)
+FULL_IMAGE = RNG.random((GRAINS.shape[0], GRAINS.shape[1]))
```
Comment (bobturneruk) [inline tests/conftest.py:417]: Should this be a fixture itself? Maybe not.

File: tests/test_dnatracing.py (hunk 1 above)
Comment (bobturneruk) [inline tests/test_dnatracing.py:8]: Should this be a fixture itself? Maybe not.

File: topostats/run_topostats.py [hunk 4]
```diff
@@ -176,16 +176,16 @@ def process_scan(
     dna_traces = dnaTrace(
         full_image_data=grains.images["gaussian_filtered"].T,
         grains=grains.images["labelled_regions"],
-        afm_image_name=filtered_image.filename,
+        filename=filtered_image.filename,
         pixel_size=filtered_image.pixel_to_nm_scaling,
-        number_of_columns=grains.images["labelled_regions"].shape[0],
-        number_of_rows=grains.images["labelled_regions"].shape[1],
     )
+    dna_traces.trace_dna()
```
Comment (bobturneruk) [inline topostats/run_topostats.py:182]: Sorry if I've asked this before, but should it be a configurable option to run this? Maybe a future decsision.

File: tests/test_dnatracing.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+                [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
+                [0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0],
+            ]
+        ),
+    }
+    for x in range(1, 4):
+        np.testing.assert_array_equal(test_dnatracing._get_grain_array(x), TARGET[x])
```
Comment (bobturneruk) [inline tests/test_dnatracing.py:63]: I've been outPythoned by this line! Maybe this isn't the forum, but it looks to me like `_get_grain_array` is a "private member of `dnatracing`, yet here it is apparently being called from `test_dnatracing`. Seems to work, but I'm curious!

Comment (bobturneruk) [reply tests/test_dnatracing.py:63]: Ah, I get it now. There's an instance of something called `test_dnatracing`. I was confused by the filename. Silly me!

**Review by bobturneruk — APPROVED** [review]
Looks good and the initial refactoring seems to be going in a good direction.

File: tests/test_dnatracing.py (hunk 1 above)
Comment (ns-rse) [reply tests/test_dnatracing.py:8]: In `tests/test_dnatracing.py` I'm focusing on isolated testing of each method so there shouldn't be any need for each to be a `@pytest.fixture`.

However, I've setup a concurrent `tests/test_dnatracing_minicircle.py` which will take and test the `@pytest.fixture`'s defined in `tests/conftest.py` where there will be a "cascading" trail of fixtures, each built on the previous fixture. At the moment I've not started that cascade (although could have included one for `_get_grain_array()`. I'll make sure I add it in the next PR.

File: tests/conftest.py (hunk 3 above)
Comment (ns-rse) [reply tests/conftest.py:417]: No, `@pytest.fixture` apply to functions, whilst `GRAINS` is a constant hence it being upper case.

File: tests/conftest.py (hunk 2 above)
Comment (ns-rse) [reply tests/conftest.py:404]: No, `@pytest.fixture` apply to functions, whilst `GRAINS` is a constant hence it being upper case.

File: tests/test_dnatracing.py (hunk 5 above)
Comment (ns-rse) [reply tests/test_dnatracing.py:63]: Not the clearest of naming conventions, sorry.

You've clocked it correctly though, `_get_grain_array()` is indeed a private method of the `dnaTracing` class that takes one argument, the number of the grain, and returns an array of 0/1 for that grain alone. This loop therefore runs over all the grains and checks the correct array is returned for each grain.

File: tests/test_dnatracing.py (hunk 1 above)
Comment (ns-rse) [reply tests/test_dnatracing.py:8]: Think this is the same as comment above, as these are unit-tests testing methods in isolation then I don't envisage needing them to be fixtures. If fixtures are needed I'll define them in `tests/conftest.py`.

File: topostats/run_topostats.py (hunk 4 above)
Comment (ns-rse) [reply topostats/run_topostats.py:182]: Its a good point and we should look at making the end-point of processing optional, I've documented it as an issue to be addressed (see #185 ).

**Pull request merged**